test/integration/api/datasets.js

```diff
@@ -37,3 +38,68 @@ describe('projects/:id/datasets GET', () => {
                 ]);
               }))))));
 });
+
+describe('projects/:id/forms/:formId/draft/attachment/link-dataset POST', () => {
+  it('should link dataset to form and returns in manifest', testService((service) =>
+    service.login('alice', (asAlice) =>
+      asAlice.post('/v1/projects/1/forms')
+        .send(testData.forms.withAttachments)
+        .set('Content-Type', 'application/xml')
+        .expect(200)
+        .then(() => asAlice.post('/v1/projects/1/forms?publish=true')
+          .send(testData.forms.simpleEntity.replace(/people/, 'goodone')))
+        .then(() => asAlice.post('/v1/projects/1/forms/withAttachments/draft/attachments/goodone.csv/link-dataset')
+          .expect(200))
+        .then(() => asAlice.post('/v1/projects/1/forms/withAttachments/draft/publish')
+          .expect(200))
+        .then(() => asAlice.get('/v1/projects/1/forms/withAttachments/manifest')
+          .set('X-OpenRosa-Version', '1.0')
+          .expect(200)
+          .then(({ text }) => {
+            const domain = config.get('default.env.domain');
+            text.should.equal(`<?xml version="1.0" encoding="UTF-8"?>
+  <manifest xmlns="http://openrosa.org/xforms/xformsManifest">
+    <mediaFile>
+      <filename>goodone.csv</filename>
+      <hash>md5:0c0fb6b2ee7dbb235035f7f6fdcfe8fb</hash>
+      <downloadUrl>${domain}/v1/projects/1/forms/withAttachments/attachments/goodone.csv</downloadUrl>
+    </mediaFile>
+  </manifest>`);
+          })))));
+
+  it('should match dataset name with attachment name', testService((service) =>
+    service.login('alice', (asAlice) =>
+      asAlice.post('/v1/projects/1/forms')
+        .send(testData.forms.withAttachments)
+        .set('Content-Type', 'application/xml')
+        .expect(200)
+        .then(() => asAlice.post('/v1/projects/1/forms?publish=true')
+          .send(testData.forms.simpleEntity))
+        .then(() => asAlice.post('/v1/projects/1/forms/withAttachments/draft/attachments/goodone.csv/link-dataset')
```

Comment:
The invalid parameter problem above where blob id and dataset id can't both be set, is there a way to trigger that through the API? 

Could/would it happen when someone defines `goodone.csv` in the form, attaches a real CSV attachment, and then links it with a dataset?

Separately, in that case, would we replace the CSV with the dataset link? What happens to the CSV - does it get lost?

Reply:
Invalid parameter problem can never occur through API, only by developers mistake. I have added integration test to ensure that never happens design time as well